app/src/main/java/edu/mills/heartoftaiwanese/activity/home/HomeFragment.kt

```diff
@@ -0,0 +1,213 @@
+            binding.submitCh.visibility = View.VISIBLE
+            binding.editTextCh.visibility = View.VISIBLE
+        }
+        if (currentTab == 0) {
+            binding.submitEn.visibility = View.VISIBLE
+            binding.editTextEng.visibility = View.VISIBLE
+        }
+        binding.clearButton.visibility = View.VISIBLE
+    }
+
+    override fun onChineseFetched(chinese: String) {
+        viewModel.fetchTaiwanese(chinese)
+    }
+
+    override fun onTaiwaneseFetched(taiwanese: String) {
+        binding.result.text = taiwanese
+        activity?.runOnUiThread {
+            showAfterSubmit()
+        }
+    }
+
+    override fun onNetworkError(error: WebResultCode) {
+        Log.e("HomeFragment", error.toString())
+        activity?.runOnUiThread {
+            showAfterSubmit()
+            when (error) {
+                WebResultCode.RATE_LIMITED -> Toast.makeText(
+                    activity,
+                    getText(R.string.too_many_requests),
+                    Toast.LENGTH_LONG
+                ).show()
+                WebResultCode.INVALID_NOT_FOUND -> Toast.makeText(
+                    activity,
+                    getText(R.string.errorNotFound),
+                    Toast.LENGTH_LONG
+                ).show()
+                WebResultCode.UNKNOWN_ERROR -> Toast.makeText(
+                    activity,
+                    getText(R.string.error),
+                    Toast.LENGTH_LONG
+                ).show()
+                // Bad state
+                WebResultCode.RESULT_OK -> throw IllegalStateException("Do not call error when there's no error")
+            }
+        }
+    }
+
+    /**
+     * Called when a tab enters the selected state.
+     *
+     * @param tab The tab that was selected
+     */
+    override fun onTabSelected(tab: TabLayout.Tab?) {
+        currentTab = tab?.position
+        if (currentTab == 1) {
+            binding.editTextEng.visibility = View.GONE
+            binding.submitEn.visibility = View.GONE
+        }
+        if (currentTab == 0) {
+            binding.editTextCh.visibility = View.GONE
+            binding.submitCh.visibility = View.GONE
+        }
+    }
+
+    /**
+     * Called when a tab exits the selected state.
+     *
+     * @param tab The tab that was unselected
+     */
+    override fun onTabUnselected(tab: TabLayout.Tab?) {
+        if (currentTab == 1) {
+            binding.editTextEng.visibility = View.VISIBLE
+            binding.submitEn.visibility = View.VISIBLE
+        }
+        if (currentTab == 0) {
+            binding.editTextCh.visibility = View.VISIBLE
+            binding.submitCh.visibility = View.VISIBLE
+        }
+    }
+
+    /**
+     * Called when a tab that is already selected is chosen again by the user. Some applications may
+     * use this action to return to the top level of a category.
+     *
+     * @param tab The tab that was reselected.
+     */
+    override fun onTabReselected(tab: TabLayout.Tab?) {
```

Comment:
Why do you have this?

Reply:
The tab selection listener requires both to be implemented.